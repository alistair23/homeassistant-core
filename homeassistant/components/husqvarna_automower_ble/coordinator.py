"""Provides the DataUpdateCoordinator."""

from __future__ import annotations

from datetime import timedelta
import logging
from typing import Any

from automower_ble.mower import Mower

from homeassistant.components import bluetooth
from homeassistant.core import HomeAssistant
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.update_coordinator import (
    CoordinatorEntity,
    DataUpdateCoordinator,
)

_LOGGER = logging.getLogger(__name__)

SCAN_INTERVAL = timedelta(seconds=60)


class Coordinator(DataUpdateCoordinator[dict[str, bytes]]):
    """Class to manage fetching data."""

    def __init__(
        self,
        hass: HomeAssistant,
        logger: logging.Logger,
        mower: Mower,
        device_info: DeviceInfo,
        address: str,
        model: str,
    ) -> None:
        """Initialize global data updater."""
        super().__init__(
            hass=hass,
            logger=logger,
            name="Husqvarna Automower BLE Data Update Coordinator",
            update_interval=SCAN_INTERVAL,
        )
        self.address = address
        self.model = model
        self.mower = mower
        self.device_info = device_info

    async def async_shutdown(self) -> None:
        """Shutdown coordinator and any connection."""
        _LOGGER.debug("Shutdown")
        await super().async_shutdown()
        if self.mower.is_connected():
            await self.mower.disconnect()

    async def _async_update_data(self) -> dict[str, bytes]:
        """Poll the device."""
        _LOGGER.debug("Polling device")

        data: dict[str, bytes] = {}

        if await self.mower.is_connected():
            device = bluetooth.async_ble_device_from_address(
                self.hass, self.address, connectable=True
            )
            if await self.mower.connect(device) == False:
                return

        data["battery_level"] = await self.mower.battery_level()
        _LOGGER.debug(data["battery_level"])
        data["activity"] = await self.mower.mower_activity()
        _LOGGER.debug(data["activity"])
        data["state"] = await self.mower.mower_state()
        _LOGGER.debug(data["state"])

        return data


class HusqvarnaAutomowerBleEntity(CoordinatorEntity[Coordinator]):
    """Coordinator entity for Husqvarna Automower Bluetooth."""

    _attr_has_entity_name = True

    def __init__(self, coordinator: Coordinator, context: Any = None) -> None:
        """Initialize coordinator entity."""
        super().__init__(coordinator, context)
        self._attr_device_info = coordinator.device_info

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return self.coordinator.mower.is_connected()
