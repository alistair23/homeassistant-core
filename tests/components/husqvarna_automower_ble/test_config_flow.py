"""Test the Husqvarna Bluetooth config flow."""

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant import config_entries
from homeassistant.components.husqvarna_automower_ble.const import DOMAIN
from homeassistant.core import HomeAssistant

from . import (
    AUTOMOWER_MISSING_SERVICE_SERVICE_INFO,
    AUTOMOWER_SERVICE_INFO,
    AUTOMOWER_UNNAMED_SERVICE_INFO,
    AUTOMOWER_UNSUPPORTED_GROUP_SERVICE_INFO,
)

from tests.components.bluetooth import inject_bluetooth_service_info

pytestmark = pytest.mark.usefixtures("mock_setup_entry")


async def test_user_selection(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test we can select a device."""

    inject_bluetooth_service_info(hass, AUTOMOWER_SERVICE_INFO)
    inject_bluetooth_service_info(hass, AUTOMOWER_UNNAMED_SERVICE_INFO)

    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )
    assert result == snapshot

    result = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={"address": "00000000-0000-0000-0000-000000000001"},
    )
    assert result == snapshot

    result = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={},
    )


async def test_no_devices(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test missing device."""

    inject_bluetooth_service_info(hass, AUTOMOWER_MISSING_SERVICE_SERVICE_INFO)
    inject_bluetooth_service_info(hass, AUTOMOWER_UNSUPPORTED_GROUP_SERVICE_INFO)

    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )
    assert result == snapshot


async def test_bluetooth(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test bluetooth device discovery."""

    inject_bluetooth_service_info(hass, AUTOMOWER_SERVICE_INFO)

    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": config_entries.SOURCE_BLUETOOTH},
        data=AUTOMOWER_SERVICE_INFO,
    )
    assert result == snapshot

    result = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        user_input={},
    )


async def test_bluetooth_invalid(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
) -> None:
    """Test bluetooth device discovery with invalid data."""

    inject_bluetooth_service_info(hass, AUTOMOWER_UNSUPPORTED_GROUP_SERVICE_INFO)

    result = await hass.config_entries.flow.async_init(
        DOMAIN,
        context={"source": config_entries.SOURCE_BLUETOOTH},
        data=AUTOMOWER_UNSUPPORTED_GROUP_SERVICE_INFO,
    )
    assert result == snapshot
