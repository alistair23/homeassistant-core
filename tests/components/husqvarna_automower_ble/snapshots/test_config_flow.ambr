# serializer version: 1
# name: test_bluetooth
  FlowResultSnapshot({
    'data_schema': None,
    'description_placeholders': dict({
      'name': 'Husqvarna Automower',
    }),
    'errors': None,
    'flow_id': <ANY>,
    'handler': 'husqvarna_automower_ble',
    'last_step': None,
    'preview': None,
    'step_id': 'confirm',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_bluetooth.1
  FlowResultSnapshot({
    'context': dict({
      'confirm_only': True,
      'source': 'bluetooth',
      'title_placeholders': dict({
        'name': 'Husqvarna Automower',
      }),
      'unique_id': '00000000-0000-0000-0000-000000000001',
    }),
    'data': dict({
      'address': '00000000-0000-0000-0000-000000000001',
      'client_id': <ANY>,
    }),
    'description': None,
    'description_placeholders': None,
    'flow_id': <ANY>,
    'handler': 'husqvarna_automower_ble',
    'minor_version': 1,
    'options': dict({
    }),
    'result': ConfigEntrySnapshot({
      'data': dict({
        'address': '00000000-0000-0000-0000-000000000001',
        'client_id': <ANY>,
      }),
      'disabled_by': None,
      'domain': 'husqvarna_automower_ble',
      'entry_id': <ANY>,
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'bluetooth',
      'title': 'Husqvarna Automower',
      'unique_id': '00000000-0000-0000-0000-000000000001',
      'version': 1,
    }),
    'title': 'Husqvarna Automower',
    'type': <FlowResultType.CREATE_ENTRY: 'create_entry'>,
    'version': 1,
  })
# ---
# name: test_bluetooth_invalid
  FlowResultSnapshot({
    'data_schema': None,
    'description_placeholders': dict({
      'name': 'Husqvarna Automower',
    }),
    'errors': None,
    'flow_id': <ANY>,
    'handler': 'husqvarna_automower_ble',
    'last_step': None,
    'preview': None,
    'step_id': 'confirm',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_no_devices
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'address',
        'required': True,
        'type': 'string',
      }),
    ]),
    'description_placeholders': None,
    'errors': None,
    'flow_id': <ANY>,
    'handler': 'husqvarna_automower_ble',
    'last_step': None,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_user_selection
  FlowResultSnapshot({
    'data_schema': list([
      dict({
        'name': 'address',
        'required': True,
        'type': 'string',
      }),
    ]),
    'description_placeholders': None,
    'errors': None,
    'flow_id': <ANY>,
    'handler': 'husqvarna_automower_ble',
    'last_step': None,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_user_selection.1
  FlowResultSnapshot({
    'data_schema': None,
    'description_placeholders': dict({
      'name': 'Husqvarna Automower',
    }),
    'errors': None,
    'flow_id': <ANY>,
    'handler': 'husqvarna_automower_ble',
    'last_step': None,
    'preview': None,
    'step_id': 'confirm',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_user_selection.2
  FlowResultSnapshot({
    'context': dict({
      'confirm_only': True,
      'source': 'user',
      'title_placeholders': dict({
        'name': 'Husqvarna Automower',
      }),
      'unique_id': '00000000-0000-0000-0000-000000000001',
    }),
    'data': dict({
      'address': '00000000-0000-0000-0000-000000000001',
      'client_id': <ANY>,
    }),
    'description': None,
    'description_placeholders': None,
    'flow_id': <ANY>,
    'handler': 'husqvarna_automower_ble',
    'minor_version': 1,
    'options': dict({
    }),
    'result': ConfigEntrySnapshot({
      'data': dict({
        'address': '00000000-0000-0000-0000-000000000001',
        'client_id': <ANY>,
      }),
      'disabled_by': None,
      'domain': 'husqvarna_automower_ble',
      'entry_id': <ANY>,
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'title': 'Husqvarna Automower',
      'unique_id': '00000000-0000-0000-0000-000000000001',
      'version': 1,
    }),
    'title': 'Husqvarna Automower',
    'type': <FlowResultType.CREATE_ENTRY: 'create_entry'>,
    'version': 1,
  })
# ---
